### DISCUSSION TITLE
how can I set a new stream-ordered memory rescource to cuco:static_map?

### DISCUSSION THREAD

**Discussion by Autumn1998** [NVIDIA/cuCollections#399] [Q&A] [answered]
I found the consturctor of static_map at source code, which accept 5 parameters as described:
![0af4cecd0e8f4d8dd6a2acdde4e375f](https://github.com/NVIDIA/cuCollections/assets/30105323/b4d0f2b7-aeac-41d1-a46e-7fe45a09829a)
I also find the default cuco:cuda_allocator(https://github.com/NVIDIA/cuCollections/blob/dev/include/cuco/utility/allocator.hpp#L27-L73), so I wrote a new allocator like cuco:cuda_allocator just added a parameter to initialize cudaStream.
but I still got  
`error: no instance of constructor "cuc(::static _map<key, Value, Scope,Allocator>::static_map [with Key=monad::content::DocId, Value=int, Scope=cuda::std::..4.._detai::thread_scope_device, Allocator=cuco::cuda allocator<char>]" matches the argument listarpument types are: 
(const std::size t, cuco:sentinel::empty keymonad::content::DocIo>, cuco::sentinel::empty_value<int>，dairy_queen::Allocator<char>，cudaStream t)cuco::static map<monad::content::DocId, int> mapf`
In fact. the consturctor only work when I set the allocator parameter as "{}". 

How can I make this parameter work?  I want to use rapidas:rmm as a general memory manager.

**Comment by Autumn1998** [comment]
I solved this problem. 
anyway, another question:  if I want to use my own allocator<T> to a static_map<Key=double, value=int>, how can I decide the type of T?
as the function provided by cuco::cuda_allocator, the allocator<T> will allocate/deallocate "n" elements. The parameter converted to allocate/deallocate is a size_t value: n. The space size is computed by sizeof(T)*n.
The answer maybe one of (T = cuco:pair<Key=double, value=int>) and (T = char). I think the former is right because the sourced code in kernels is set "capacity" as n. but I found that at some test code(tests/static_map/insert_or_assign_test.cu etc...), the cuda_allocator<char> is used.

**Reply by PointKernel** [reply, marked as answer]
Ideally, the map slot type is desired, e.g., [cuco::pair<Key, T>](https://github.com/NVIDIA/cuCollections/blob/dev/include/cuco/static_map.cuh#L96), but we are using [allocator traits internally converting to the proper type](https://github.com/NVIDIA/cuCollections/blob/dev/include/cuco/aow_storage.cuh#L69-L70) so what's passed by the user doesn't really matter.

**Reply by PointKernel** [reply]
Another note: you may want to use the new `cuco::experimental::static_map` instead of the legacy `cuco::static_map` since the latter will be deprecated soon.

**Reply by Autumn1998** [reply]
I noticed the "insert_and_find" can only be applied when CG_SIZE =1 , but this may caused low performance problem when the datasize>1M?

**Reply by sleeepyjack** [reply]
IIRC `cuco::experimental::static_map::insert_and_find` works with any CG size.
A larger CG size is beneficial if the map's occupancy exceeds a certain threahold, e.g. 50% load factor or so.

**Comment by PointKernel** [comment]
> How can I make this parameter work? I want to use rapidas:rmm as a general memory manager.

You may want to pass your allocator to `stream_allocator_adaptor` (https://github.com/rapidsai/rmm/blob/branch-24.02/include/rmm/mr/device/polymorphic_allocator.hpp#L149) and pass the adaptor as an allocator to cuco maps.

**Reply by Autumn1998** [reply]
thanks for your kindly reply that helps a lot
